File: backend-spring/src/main/java/com/codequest/controller/LevelProgressController.java
package com.codequest.controller;

import com.codequest.dto.LevelProgressDto;
import com.codequest.dto.UpdateLevelProgressRequest;
import com.codequest.service.LevelProgressService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/levels")
@Tag(name = "Level Progress")
@SecurityRequirement(name = "bearerAuth")
public class LevelProgressController {

    private final LevelProgressService levelProgressService;

    public LevelProgressController(LevelProgressService levelProgressService) {
        this.levelProgressService = levelProgressService;
    }

    @GetMapping("/progress")
    public ResponseEntity<?> getLevelProgress(
            @RequestParam String language,
            @RequestParam String track,
            @RequestParam Integer levelId) {
        Optional<LevelProgressDto> progress = levelProgressService.getLevelProgress(language, track, levelId);
        return ResponseEntity.ok(progress.orElse(null));
    }

    @GetMapping("/progress/all")
    @Operation(summary = "Get all progress for current user")
    public ResponseEntity<List<LevelProgressDto>> getAllProgress() {
        return ResponseEntity.ok(levelProgressService.getAllProgress());
    }

    @PutMapping("/progress")
    public ResponseEntity<LevelProgressDto> updateLevelProgress(@RequestBody UpdateLevelProgressRequest request) {
        return ResponseEntity.ok(levelProgressService.updateLevelProgress(request));
    }
}

File: backend-spring/src/main/kotlin/com/codequest/controller/LevelProgressController.kt
package com.codequest.controller

import com.codequest.dto.LevelProgressDto
import com.codequest.dto.UpdateLevelProgressRequest
import com.codequest.service.LevelProgressService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.security.SecurityRequirement
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/levels")
@Tag(name = "Level Progress", description = "Level progress tracking")
@SecurityRequirement(name = "bearerAuth")
class LevelProgressController(
    private val levelProgressService: LevelProgressService
) {

    @GetMapping("/progress")
    @Operation(summary = "Get level progress", description = "Returns progress for a specific level")
    fun getLevelProgress(
        @RequestParam language: String,
        @RequestParam track: String,
        @RequestParam levelId: Int
    ): ResponseEntity<LevelProgressDto?> {
        return ResponseEntity.ok(levelProgressService.getLevelProgress(language, track, levelId))
    }

    @PutMapping("/progress")
    @Operation(summary = "Update level progress", description = "Saves or updates level progress (upsert)")
    fun updateLevelProgress(@RequestBody request: UpdateLevelProgressRequest): ResponseEntity<LevelProgressDto> {
        return ResponseEntity.ok(levelProgressService.updateLevelProgress(request))
    }
}

